### PR TITLE
modern lstm

### PR DESCRIPTION
**Pull request by wangjiawen2013**
### Checklist

- [ ] Confirmed that `run-checks all` script has been executed.
- [ ] Made sure the book is up to date with changes in this PR.

### Related Issues/PRs

_Provide links to relevant issues and dependent PRs._

### Changes

_Summarize the problem being addressed and your solution._

### Testing

_Describe how these changes have been tested._

### CODE REVIEW

**Comment by wangjiawen2013** [comment]
Here is a test result, the results maybe a little different each time because the datasets were randomly generated.



This implementation is complementary to Burn's official LSTM, users can choose either one depends on the project's specific needs.

File: examples/modern-lstm/README.md [hunk 1]
```diff
@@ -0,0 +1,22 @@
+# Advanced LSTM Implementation with Burn
+A sophisticated implementation of Long Short-Term Memory (LSTM) networks in Burn, featuring state-of-the-art architectural enhancements and optimizations. This implementation includes bidirectional processing capabilities and advanced regularization techniques, making it suitable for both research and production environments. More details can be found at the [PyTorch implementation](https://github.com/shiv08/Advanced-LSTM-Implementation-with-PyTorch).
```
Comment (laggui) [inline examples/modern-lstm/README.md:2]: I understand this is taken from the reference implementation README, but this description is a bit exaggerated 😅 I would probably change the content of the paragraph.

It is an LSTM implementation that concatenates the weights for the input and hidden transformations (for efficiency, similar to the official PyTorch layer implementation) and adds layer norm & dropout in the mix.

Performing four (forget, input, output and cell gates) matrix multiplications or concatenating the weights to perform one bigger matrix multiplication and separate the results afterwards should have the same results. It should help especially with smaller inputs to better leverage the GPU.

File: examples/modern-lstm/README.md [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+## Example Usage
+
+### Training
+```sh
+cargo run --release --features ndarray -- train --artifact-dir /home/wangjw/data/work/projects/lstm/output --num-epochs 30 --batch-size 32 --num-workers 2 --lr 0.001
+```
+
+### Inference
+```sh
+cargo run --release --features ndarray -- infer --artifact-dir /home/wangjw/data/work/projects/lstm/output
```
Comment (laggui) [inline examples/modern-lstm/README.md:21]: We should change the value for the example artifact dir 😄 could be `/tmp/modern-lstm`

File: examples/modern-lstm/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+                    lr,
+                );
+                train::<Autodiff<NdArray>>(&artifact_dir, config, device);
+            }
+            Commands::Infer { artifact_dir } => {
+                infer::<NdArray>(&artifact_dir, device);
+            }
+        }
+    }
+}
```
Comment (laggui) [inline examples/modern-lstm/src/main.rs:50]: Instead of duplicating the content of `run` for each backend feature flag, you can keep it generic on the backend and have it take a device argument. Then, you can call it with different a different backend & device instead. Will reduce code duplication.

File: examples/modern-lstm/src/model.rs [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+// g_t = tanh(W_g · [h_{t-1}, x_t] + b_g]   # Candidate cell state
+// o_t = σ(W_o · [h_{t-1}, x_t] + b_o)      # Output gate
+//
+// c_t = f_t ⊙ c_{t-1} + i_t ⊙ g_t            # New cell state
+// h_t = o_t ⊙ tanh(c_t)                       # New hidden state
+//
+// where:
+// - σ is the sigmoid function
+// - ⊙ is the element-wise multiplication
+// - [h_{t-1}, x_t] represents concatenation
```
Comment (laggui) [inline examples/modern-lstm/src/model.rs:23]: Docstrings take 3 forward slashes `///`. This also applies to the other docstrings in this file.

File: examples/modern-lstm/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+rand_distr = { version="0.4.3" }
+
+# Serialization
+serde = { version="1.0.210", features=["std", "derive"] }
+
+# Organise the results in dataframe
+polars = { version="0.44.1" }
+
+# Command line parser
+clap = { version="4.5.21", features=["derive"] }
```
Comment (laggui) [inline examples/modern-lstm/Cargo.toml:29]: We should use existing workspace dependencies.
```suggestion
burn = { path = "../../crates/burn", features=["train"] }

# Random number generator
rand = { workspace = true }
rand_distr = { workspace = true }

# Serialization
serde = {workspace = true, features = ["std", "derive"]}

# Organise the results in dataframe
polars = { workspace = true }

# Command line parser
clap = { workspace = true }
```

**Review by laggui — CHANGES_REQUESTED** [review]
Great example! 🙂 I have some minor changes in mind.

We should also add it to the list of examples in the [README](https://github.com/tracel-ai/burn/blob/main/README.md?plain=1#L543) and [book](https://github.com/tracel-ai/burn/blob/main/burn-book/src/examples.md?plain=1#L72).

The example could be restructured to follow the workspace examples (similar to what I did in your last wgan example), but that's not mandatory.

**Comment by wangjiawen2013** [comment]
Good! I am not so familiar with Rust and Burn as you. I am glad that you can revise the code. I'll learn from you on the design logic and formatting.

**Comment by wangjiawen2013** [comment]
Good! I am not so familiar with Rust and Burn as you. I am glad that you can revise the code. I'll learn from you on the design logic and formatting.
I am not clear about the working mechanisms of github. I find that you have modified the code in my pull request. Do I need modify the source code and pull a request again, or you will make all the changes you mentioned instead of me ? If I need to make the changes you suggest, I will modify and test it in my fork and pull a request again.

**Comment by laggui** [comment]
> I am not clear about the working mechanisms of github. I find that you have modified the code in my pull request.

The specific changes for the `Cargo.toml` were pretty short so I was able to suggest the code changes from the github PR review, and as you can see it automatically committed them to your fork when you applied them.

> Do I need modify the source code and pull a request again, or you will make all the changes you mentioned instead of me ? If I need to make the changes you suggest, I will modify and test it in my fork and pull a request again.

The contribution process usually happens in a sequential manner. A user (you) submits code changes, a contributor (me) reviews the code and requests modifications where needed, and the user can provide his thoughts on the requests and make the necessary changes before asking for another round of review. And the process is typically repeated. But everything remains as part of the same pull request 🙂 

In your last PR, the only changes required were structural (change how the example needed to be exposed in the workspace) so I made the changes myself since I had time.

File: examples/modern-lstm/README.md (hunk 1 above)
Comment (wangjiawen2013) [reply examples/modern-lstm/README.md:2]: You are a native English speaker and professional deep learning expert and of course you can change the Introduction if it is exaggerated. 

Is it reasonable to use weights concatenation for LSTM ?

I have try my best to revise the code according to your suggestions. I tested it using `ndarray` because we have no `GPU` in our server and it worked as expected. There maybe some insufficiencies on docstrings because I don't know which part should be docstrings, if you find them, you can modified them.

**Comment by wangjiawen2013** [comment]
It's weired that the model performance decreased a little after these changes. Is the latest burn  still unstable or the crates in the `Cargo.toml` will affect the performance ? I didn't change any parameters and the model. The only things changed are `Cargo.toml` and the shell commands. I tested the orginal one again, it indeed worked better than this new one.
This is the new one:
new

**Review by laggui — APPROVED** [review]
Fixed some conflicts with the main branch.

Thanks for the great example 🙂 we might add this LSTM implementation to our actual modules in the future
